src/rewriter.ts

```diff
@@ -126,26 +136,44 @@ export abstract class Rewriter {
     if (from >= fullStart && from < textStart) {
       from = Math.max(from, this.skipCommentsUpToOffset);
     }
-    // Add a source mapping. writeRange(from, to) always corresponds to
-    // original source code, so add a mapping at the current location that
-    // points back to the location at `from`. The additional code generated
-    // by tsickle will then be considered part of the last mapped code
-    // section preceding it. That's arguably incorrect (e.g. for the fake
-    // methods defining properties), but is good enough for stack traces.
-    const pos = this.file.getLineAndCharacterOfPosition(from);
-    this.sourceMapper.addMapping(
-        node, {line: pos.line, column: pos.character, position: from}, this.position, to - from);
     // getSourceFile().getText() is wrong here because it has the text of
     // the SourceFile node of the AST, which doesn't contain the comments
     // preceding that node.  Semantically these ranges are just offsets
     // into the original source file text, so slice from that.
     const text = this.file.text.slice(from, to);
-    if (text) {
-      this.emit(text);
-    }
+    const pos = this.file.getLineAndCharacterOfPosition(from);
+    // Source maps are line based, so if str has leading whitespace, particularly
+    // newlines, we want to map to the start of/line of the actual text
+    const originalSourcePosition = this.findOffsetOfFirstNonWhitespaceCharacter(
+        text, {line: pos.line, column: pos.character, position: from});
+    this.emit(text, node, originalSourcePosition);
   }
 
-  emit(str: string) {
+  /**
+   * @param str string to emit
+   * @param node the node to source map str to
+   * @param original the position in the source file to source map str to
+   *
+   * Every emit call results in a source mapping being made, if a node and
+   * original SourcePosition aren't supplied, str is source mapped to the
+   * beginning of the current node.
+   */
+  emit(str: string): void;
+  emit(str: string, node: ts.Node, original: SourcePosition): void;
+  emit(str: string, node?: ts.Node, original?: SourcePosition): void {
+    if (!node || !original) {
+      // Only the transformer source mapper uses the node, and providing
+      // a source mapping for every emit messes up the transformer
+      // code path, so unless we're emitting copied text from a node, don't
+      // set it
+      original = this.sourcePositionOfNodeStart(this.getCurrentNode());
```

Comment:
To make sourcemaps work well with transformers, we should still pass the `this.getCurrentNode()` to `addMapping`, however with an additional flag that controls whether the transformer does a `ts.setOriginalNode` or just a `ts.setSourceMapRange`...
I think this difference is the problem that you ran into, as right now we are always doing both.